src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -1050,197 +1036,265 @@ void ImageReplayer<I>::shut_down_journal_replay(bool cancel_ops)
-      if (m_last_r < 0) {
-	status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_ERROR;
-	status.description = m_state_desc;
-      } else {
-	status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPED;
-	status.description = m_state_desc.empty() ? "stopped" : m_state_desc;
+      std::string desc;
+      if (!m_replay_status_formatter->get_or_send_update(&desc,
+                                                         on_req_finish)) {
+        dout(20) << "waiting for replay status" << dendl;
+        return;
       }
+      status.description = "replaying, " + desc;
+    }
+    break;
+  case STATE_STOPPING:
+    if (stopping_replay) {
+      status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPING_REPLAY;
+      status.description = "stopping replay";
       break;
-    default:
-      assert(!"invalid state");
     }
-  }
-
-  if (status.state == cls::rbd::MIRROR_IMAGE_STATUS_STATE_REPLAYING) {
-    Context *on_req_finish = new FunctionContext(
-      [this](int r) {
-	if (r == 0) {
-	  librados::AioCompletion *comp = nullptr;
-	  {
-	    Mutex::Locker locker(m_lock);
-	    std::swap(m_update_status_comp, comp);
-	  }
-	  if (comp) {
-	    comp->release();
-	  }
-	  update_mirror_image_status(false, STATE_REPLAYING);
-	}
-      });
-    std::string desc;
-    if (!m_replay_status_formatter->get_or_send_update(&desc, on_req_finish)) {
-      return;
+    // FALLTHROUGH
+  case STATE_STOPPED:
+    if (last_r < 0) {
+      status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_ERROR;
+      status.description = state_desc;
+    } else {
+      status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPED;
+      status.description = state_desc.empty() ? "stopped" : state_desc;
     }
-    status.description = "replaying, " + desc;
+    break;
+  default:
+    assert(!"invalid state");
   }
 
   dout(20) << "status=" << status << dendl;
-
   librados::ObjectWriteOperation op;
   librbd::cls_client::mirror_image_status_set(&op, m_global_image_id, status);
 
-  if (final) {
-    reschedule_update_status_task(-1);
-    m_local_ioctx.aio_flush();
-    librados::AioCompletion *comp = nullptr;
-    {
+  librados::AioCompletion *aio_comp = create_rados_ack_callback<
+    ImageReplayer<I>, &ImageReplayer<I>::handle_mirror_status_update>(this);
+  int r = m_local_ioctx.aio_operate(RBD_MIRRORING, aio_comp, &op);
+  assert(r == 0);
+  aio_comp->release();
+}
+
+template <typename I>
+void ImageReplayer<I>::handle_mirror_status_update(int r) {
+  dout(20) << "r=" << r << dendl;
+
+  bool update_status_requested = false;
+  {
+    Mutex::Locker locker(m_lock);
+    std::swap(update_status_requested, m_update_status_requested);
+  }
+
+  // if a deferred update is available, send it -- otherwise reschedule
+  // the timer task
+  if (is_running()) {
+    bool started = false;
+    if (update_status_requested) {
       Mutex::Locker locker(m_lock);
-      std::swap(m_update_status_comp, comp);
+      started = start_mirror_image_status_update(false, true);
     }
```

Comment:
I don't like very much we take (and drop) the lock 3 times here (to get update_status_requested, to check is_running, and to start_mirror_image_status_update). Although it looks like race is not possible, it is still alarming and could be avoided (take the lock only once) if use `is_running_()`?